### PR TITLE
Merge emit_small_memcpy and emit_small_memmove

### PR DESCRIPTION
**Pull request by bjorn3**
- [x] This has been discussed in issue #..., or if not, please tell us why
  here. This is mostly a cleanup.
- [x] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first. They share a lot of code. Also by performing all loads before all stores, the performance of the simple-raytracer cg_clif benchmark increases by 1%.
- [x] This PR contains test cases, if meaningful. N/A
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so and/or ping
  `bnjbvr`. <strike>The list of suggested reviewers on the right can help you.</strike> (no list for non members)

### CODE REVIEW

**Comment by bjorn3** [comment]
I found a bug with cg_clif. Will investigate if it is caused by this or a change to cg_clif itself.

File: cranelift-frontend/src/frontend.rs [hunk 1]
```diff
@@ -665,16 +672,27 @@ impl<'a> FunctionBuilder<'a> {
...
+            if non_overlapping {
+                self.call_memcpy(config, dest, src, size_value);
+            } else {
+                self.call_memmove(config, dest, src, size_value);
+            }
             return;
         }
 
         let mut flags = MemFlags::new();
         flags.set_aligned();
```
Comment (bjorn3) [inline cranelift-frontend/src/frontend.rs]: Pre-existing, but this needs to compare `{src,dest}_align` with `access_size`.

**Comment by bjorn3** [comment]
> I found a bug with cg_clif. Will investigate if it is caused by this or a change to cg_clif itself.

This was a bug in cg_clif itself.

File: cranelift-frontend/src/frontend.rs [hunk 2]
```diff
@@ -628,15 +627,23 @@ impl<'a> FunctionBuilder<'a> {
...
 
-    /// Optimised memcpy for small copies.
-    pub fn emit_small_memcpy(
+    /// Optimised memcpy or memmove for small copies.
+    ///
+    /// # Codegen safety
+    ///
+    /// The following properties must hold to prevent UB:
+    ///
+    /// * `src_align` and `dest_align` are an upper-bound on the alignment of `src` cq `dest`.
```
Comment (eqrion) [inline cranelift-frontend/src/frontend.rs]: `src cq dest`

Is `cq` a typo or am I missing an abbreviation?

Comment (bjorn3) [reply cranelift-frontend/src/frontend.rs]: It seems that `cq` is a Dutch abbreviation: https://nl.m.wikipedia.org/wiki/Casu_quo. I also used it wrong. :man_facepalming: I meant: `src_align` is the upper-bound on the alignment of `src` and `dest_align` is the upper-bound on the alignment of `dest`. I replaced it with `respectively`.

**Pull request merged**